### ISSUE TITLE
tracing all children of an already started program.

### ISSUE THREAD

**Issue by Zibri** [strace/strace#193]
How can I use strace to trace a program that already forked a few children and it is already executing before running strace -p on it?

My test program:

```
#include <stdio.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

int main(int argc, char *argv[])
{
  int f1, f2, f3;
  f1 = open("/dev/null", O_RDWR);
  f2 = open("/dev/null", O_RDWR);
  f3 = open("/dev/null", O_RDWR);

fork();
fork();

  int ck;
  do {
    ck = 0;
    ck += lseek(f1, 0, SEEK_CUR);
    ck += lseek(f2, 0, SEEK_CUR);
    ck += lseek(f3, 0, SEEK_CUR);
    printf("I'm aliveee...\n");
    sleep(15);
  } while (ck != -3);

  return 0;
}

```

If I use strace -f -p (pid of the running program) I only get what the main thread does...
I guess I should use multiple -p options to trace them all but I also think there should/could be a command line switch to automatically do so.

**Comment by esyr** [comment]
If you're interested in immediate children only, then `strace -f -p "$parent" -p "$(pgrep -P "$parent")"` should suffice; otherwise, something like `strace -f -p "$(l="$parent"; while [ -n "$l" ]; do echo $l; n=""; for i in $(echo "$l"); do; n="$n $(pgrep -P "$i")"; done; l="$n";  done)"` is needed.  I don't think the need to use a shell one-liner to build a list of processes of interest justifies tens and possibly hundreds of lines of code of finicky `procfs` manipulation required to implement such logic inside `strace`, especially consider all other possible PID selection needs covered by `pgrep` (same UID/GID/EUID/EGID/SUID/SGID/FSUID/FSGID/PGID/SID/PIDNS only? Direct children only? Siblings? Ancestors? How many levels?). Having an option to recursively select children in `pgrep` might be more straightforward though, considering the fact it already implements all the other possible process selection filters.

**Comment by Zibri** [comment]
> If you're interested in immediate children only, then `strace -f -p "$parent" -p "$(pgrep -P "$parent")"` should suffice; otherwise, something like `strace -f -p "$(l="$parent"; while [ -n "$l" ]; do echo $l; n=""; for i in $(echo "$l"); do; n="$n $(pgrep -P "$i")"; done; l="$n"; done)"` is needed. I don't think the need to use a shell one-liner to build a list of processes of interest justifies tens and possibly hundreds of lines of code of finicky `procfs` manipulation required to implement such logic inside `strace`, especially consider all other possible PID selection needs covered by `pgrep` (same UID/GID/EUID/EGID/SUID/SGID/FSUID/FSGID/PGID/SID/PIDNS only? Direct children only? Siblings? Ancestors? How many levels?). Having an option to recursively select children in `pgrep` might be more straightforward though, considering the fact it already implements all the other possible process selection filters.

in reality a single line is needed:
this one gets all children of a process..
`pstree -pn `pidof RELEASE` |grep -o "([[:digit:]]*)" |grep -o "[[:digit:]]*"`
so I think it's not a huge mess just to add that in the command line options..

**Comment by tryauuum** [comment]
for anyone reading this: even though `pgrep` doesn't support getting a list of all process's descendants, maybe for your use case matching by process group id will be perfect.
```
strace -f $( pgrep -g PID_OF_ANCESTOR | sed 's!^!-p !' )
```

**Comment by masatake** [comment]
> strace -f $( pgrep -g PID_OF_ANCESTOR | sed 's!^!-p !' )

This can be:
```
strace -f -p "$( pgrep -g PID_OF_ANCESTOR)"
```

**Comment by Zibri** [comment]
Yep.. I was just saying it would be a nice feature to have in the same program without using external ones.

**Comment by masatake** [comment]
> Yep.. I was just saying it would be a nice feature to have in the same program without using external ones.

I agree with you if more than one program, pstree, pidof, grep and/or sed are needed.
However, I think just one, pgrep (or pidof)  is acceptable. 
If strace introduces an option implementing the feature requested here, the user must remember the option to enable the feature.
`[tab]` key with a modern shell may help us but remembering the command fragment, `-p "$(...)"` is not hard for me.

I was directly taught this fragment by @ldv-alt, the maintainer of strace!
https://lists.strace.io/pipermail/strace-devel/2018-March/007950.html
https://lists.strace.io/pipermail/strace-devel/2018-March/007951.html

The lesson was not free. I had to submit a patch to procps-ng as fee.
https://gitlab.com/procps-ng/procps/-/merge_requests/58

**Comment by ldv-alt** [comment]
The consensus is that there is nothing that needs to be implemented on the strace side to address this, closing.
